master/internal/spec_util.go

```diff
@@ -0,0 +1,125 @@
+		&sproto.ValidateResourcePoolAvailabilityRequest{
+			Name:  poolName,
+			Slots: slots,
+		},
+	)
+	if err != nil {
+		return "", launchWarnings, fmt.Errorf("checking resource availability: %v", err.Error())
+	}
+	if a.m.config.ResourceManager.AgentRM != nil &&
+		a.m.config.LaunchError &&
+		len(launchWarnings) > 0 {
+		return "", nil, errors.New("slots requested exceeds cluster capacity")
+	}
+
+	return poolName, launchWarnings, nil
+}
+
+func fillTaskSpec(a *apiServer, poolName string, agentUserGroup *model.AgentUserGroup, userModel *model.User) (tasks.TaskSpec, error) {
+	taskContainerDefaults, err := a.m.rm.TaskContainerDefaults(
+		poolName,
+		a.m.config.TaskContainerDefaults,
+	)
+	if err != nil {
+		return tasks.TaskSpec{}, fmt.Errorf("getting TaskContainerDefaults: %v", err)
+	}
+	taskSpec := *a.m.taskSpec
+	taskSpec.TaskContainerDefaults = taskContainerDefaults
+	taskSpec.AgentUserGroup = agentUserGroup
+	taskSpec.Owner = userModel
+	return taskSpec, nil
+}
+
+func fillTaskConfig(resourcePoolDest **string, poolName string, resourceSlotsDest **int, slots int, taskSpec tasks.TaskSpec, environment *model.Environment) {
+	*resourcePoolDest = &poolName
+	*resourceSlotsDest = &slots
+
+	taskContainerPodSpec := taskSpec.TaskContainerDefaults.GPUPodSpec
+	if slots == 0 {
+		taskContainerPodSpec = taskSpec.TaskContainerDefaults.CPUPodSpec
+	}
+	environment.PodSpec = (*k8sV1.Pod)(schemas.Merge(
+		(*expconf.PodSpec)(environment.PodSpec),
+		(*expconf.PodSpec)(taskContainerPodSpec),
+	))
+}
+
+func fillContextDir(configWorkDirDest **string, defaultWorkDir *string, contextDirectory []*utilv1.File) ([]byte, error) {
+	var contextDirectoryBytes []byte
+	if len(contextDirectory) > 0 {
+		userFiles := filesToArchive(contextDirectory)
+
+		workdirSetInReq := *configWorkDirDest != nil &&
+			(defaultWorkDir == nil || *defaultWorkDir != **configWorkDirDest)
+		if workdirSetInReq {
+			return nil, status.Errorf(codes.InvalidArgument,
+				"cannot set work_dir and context directory at the same time")
+		}
+		*configWorkDirDest = nil
+
+		var err error
+		contextDirectoryBytes, err = archive.ToTarGz(userFiles)
+		if err != nil {
+			return nil, status.Errorf(codes.InvalidArgument,
+				fmt.Errorf("compressing files context files: %w", err).Error())
+		}
+	}
+	return contextDirectoryBytes, nil
+}
+
+func getTaskSessionToken(ctx context.Context, userModel *model.User) (string, error) {
+	extConfig := config.GetMasterConfig().InternalConfig.ExternalSessions
+	var token string
+	var err error
+	if extConfig.Enabled() {
+		token, err = grpcutil.GetUserExternalToken(ctx)
+		if err != nil {
+			return "", status.Errorf(codes.Internal,
+				errors.Wrapf(err,
+					"unable to get external user token").Error())
+		}
+	} else {
+		token, err = user.StartSession(ctx, userModel)
+		if err != nil {
+			return "", status.Errorf(codes.Internal,
+				errors.Wrapf(err,
+					"unable to create user session inside task").Error())
+		}
+	}
+	return token, nil
```

Comment:
couple style nits in one; var extConfig doesn't really make the code more readable and we can get away with not forward declaring token/err if we change the control flow a bit.
```suggestion
	if config.GetMasterConfig().InternalConfig.ExternalSessions.Enabled() {
		token, err = grpcutil.GetUserExternalToken(ctx)
		if err != nil {
			return "", status.Errorf(codes.Internal,
				errors.Wrapf(err,
					"unable to get external user token").Error())
		}
        return token, nil
	}

    token, err = user.StartSession(ctx, userModel)
    if err != nil {
        return "", status.Errorf(codes.Internal,
            errors.Wrapf(err,
                "unable to create user session inside task").Error())
    }
    return token, nil
```